container: refactor custom handlers interface

giuseppe opened the pull request:
some initial refactoring for the custom handlers code.

Move the entire implementation to a separate file.

Signed-off-by: Giuseppe Scrivano <gscrivan@redhat.com>

giuseppe commented:
it is just a preparatory PR, I'll iterate further and move each handler to its own file.

The goal is to be able to load them as plugins in future without hard-coding them in libcrun

flouthoc commented:
Should we move `wasm` and `libkrun` into different files or would be part of different PR.

rhatdan commented:
LGTM

flouthoc commented on src/libcrun/custom-handler.c:
```diff
@@ -0,0 +1,595 @@
+  return crun_make_error (err, ENOTSUP, "libkrun or dlopen not present");
+#endif
+}
+
```
Same
```suggestion
if HAVE_DLOPEN && HAVE_WASMER
```

giuseppe replied:
I've moved the check inside the function so the function is always defined

giuseppe commented:
added another patch to hide all the handlers behind an interface.  We might need to add more pieces before we can load handlers from a `.so`.

Moved the PR back to Draft.  I've not tried to run it yet.

flouthoc commented:
I know this is in draft. But i tried running it today for some reason handlers are not getting invoked.

giuseppe commented:
> I know this is in draft. But i tried running it today for some reason handlers are not getting invoked.

doesn't it fire when there is the wasm annotation specified?

flouthoc commented:
> > I know this is in draft. But i tried running it today for some reason handlers are not getting invoked.
> 
> doesn't it fire when there is the wasm annotation specified?

@giuseppe I have not tested it for `libkrun` yet but it does not works `wasm`. It throws `2021-12-12T11:06:23.000782755Z: open executable: Permission denied.` for `wasm` images as of now.

giuseppe commented:
I have tested libkrun and it seems to work fine.  I will test what is wrong with the wasm handlers

flouthoc commented:
Had some issues with wasmer but works perfect after this diff on top of this PR.

```diff
diff --git a/src/libcrun/container.c b/src/libcrun/container.c
index a51f018..a62bd9e 100644
--- a/src/libcrun/container.c
+++ b/src/libcrun/container.c
@@ -1142,7 +1142,7 @@ container_init_setup (void *args, pid_t own_pid, char *notify_socket,
 
   if (def->process && def->process->args)
     {
-      *exec_path = find_executable (def->process->args[0], def->process->cwd, entrypoint_args->context->handler);
+      *exec_path = find_executable (def->process->args[0], def->process->cwd, entrypoint_args->custom_handler->name);
       if (UNLIKELY (*exec_path == NULL))
         {
           if (errno == ENOENT)
diff --git a/src/libcrun/handlers/wasmedge.c b/src/libcrun/handlers/wasmedge.c
index a0a505a..91abf1b 100644
--- a/src/libcrun/handlers/wasmedge.c
+++ b/src/libcrun/handlers/wasmedge.c
@@ -178,7 +178,7 @@ wasmedge_can_handle_container (libcrun_container_t *container, libcrun_error_t *
 }
 
 struct custom_handler_s handler_wasmedge = {
-  .name = "wasmedge",
+  .name = "wasm",
   .feature_string = "WASM:wasmedge",
   .load = libwasmedge_load,
   .unload = libwasmedge_unload,
diff --git a/src/libcrun/handlers/wasmer.c b/src/libcrun/handlers/wasmer.c
index 779f23a..f6debb7 100644
--- a/src/libcrun/handlers/wasmer.c
+++ b/src/libcrun/handlers/wasmer.c
@@ -49,7 +49,6 @@ libwasmer_exec (void *cookie, libcrun_container_t *container,
   const wasm_func_t *core_func;
   FILE *wat_wasm_file;
   size_t file_size;
-  void *handle = arg;
   wasm_byte_vec_t wat;
   wasm_byte_vec_t binary_bytes;
   wasm_byte_vec_t wasm_bytes;
@@ -246,7 +245,7 @@ libwasmer_exec (void *cookie, libcrun_container_t *container,
 }
 
 static int
-libwasmer_load (struct custom_handler_s *out, libcrun_error_t *err)
+libwasmer_load (void **cookie, libcrun_error_t *err arg_unused)
 {
   void *handle;
 
@@ -285,7 +284,7 @@ libwasmer_can_handle_container (libcrun_container_t *container, libcrun_error_t
 }
 
 struct custom_handler_s handler_wasmer = {
-  .name = "wasmer",
+  .name = "wasm",
   .feature_string = "WASM:wasmer",
   .load = libwasmer_load,
   .unload = libwasmer_unload,
```

flouthoc reviewed (approved):
LGTM
